### PR TITLE
Fix matrices documentation

### PR DESCRIPTION
**Pull request by Dragorn421**
Follow-up to #923 , as promised. Only touched existing documentation, ~~I may get to documenting everything else later.~~ (I did: #955 )

@zbanks @petrie911 It would be nice if you could review this and make sure I didn't mess up somewhere

I got rid of the "roll/pitch/yaw" terminology since it just makes things more confusing imo, without bringing much.

If we agree upon using eg "XYZ" instead of "PYR" or whatever it would be, then there's more that can be fixed, for example at [z_eff_ss_dead_ds.c:99](https://github.com/zeldaret/oot/blob/master/src/overlays/effects/ovl_Effect_Ss_Dead_Ds/z_eff_ss_dead_ds.c#L99) :
```c
Matrix_RotateZYX(this->rRoll, this->rPitch, this->rYaw, MTXMODE_APPLY);
```
aka (`r***` are defines):
```c
Matrix_RotateZYX(this->regs[2], this->regs[3], this->regs[4], MTXMODE_APPLY);
```
should probably change `rRoll`, `rPitch` and `rYaw` to `rRotX`, `rRotY`, `rRotZ`

### CODE REVIEW

**Review by fig02 — APPROVED** [review]
approving that the changes overall look good, but I do not have any knowledge to properly weigh in on the content. +1 on getting some math nerds to look over this if they are able

**Comment by Dragorn421** [comment]
I noticed I used inconsistent naming

`Matrix_TranslateRotateZYX` (previously `Matrix_JointPosition`) produces `translate * rotZ * rotY * rotX` (applied right to left)
`SkinMatrix_SetRotateZYXTranslate` (previously `SkinMatrix_SetRotateRPYTranslate`) also produces `translate * rotZ * rotY * rotX`

`TranslateRotateZYX` is consistent with the matrix multiplication order,
`RotateZYXTranslate` is more consistent (apart from the ZYX) with the order the transform happens in

which one would be better to stick to, I would think the first one but not sure

File: src/code/z_skin_matrix.c [hunk 1]
```diff
@@ -224,8 +230,7 @@ void SkinMatrix_MtxFCopy(MtxF* src, MtxF* dest) {
 }
 
 /**
- * Inverts a matrix using a slight modification of the Gauss-Jordan method
- * (column operations instead of row operations).
+ * Inverts a matrix using the Gauss-Jordan method.
```
Comment (zbanks) [inline src/code/z_skin_matrix.c:233]: Great find, makes it much simpler.

File: src/code/sys_matrix.c [hunk 2]
```diff
@@ -293,12 +293,12 @@ void Matrix_RotateZ(f32 z, u8 mode) {
...
 
 /**
- * Rotates the top of the matrix stack by `z` degrees, then
- * rotates that matrix by `y` degrees, then rotates that matrix
- * by `x` degrees. (roll-pitch-yaw)
+ * Rotate using ZYX Tait-Bryan angles.
+ * This means a (column) vector is first rotated around X, then around Y, then around Z, then (if `mode` is
+ * `MTXMODE_APPLY`) gets transformed according to whatever the matrix was before adding the ZYX rotation.
  * Original Name: Matrix_RotateXYZ, changed to reflect rotation order.
  */
```
[the remark this answers is not in the document]
Comment (Dragorn421) [reply src/code/sys_matrix.c:300]: what order is backwards? `ZYX` or "first rotated around X, then around Y, then around Z" ? in the sentence I'm reading the multiplication with a vector right-to-left

[the remark this answers is not in the document]
Comment (zbanks) [reply src/code/sys_matrix.c:300]: Sorry for the ambiguity -- I think it should be called `ZYX` and the comment should say "first rotated around Z, then around Y, then around X." 

If it were to be decomposed into 3 axis-aligned rotation matrices, you're right it would be `M_x * M_y * M_z`. But, the most common convention I've seen is that the axes are in the left-to-right order they are applied. 

I think this is also relevant to your comment about the functions with `Translate`: I think `RotateZYXTranslate` is the best name, because the translate happens *after* the rotations.

[the remark this answers is not in the document]
Comment (Dragorn421) [reply src/code/sys_matrix.c:300]: But with a `ZYX` rotation matrix `rotZ * rotY * rotX` when multiplying a column vector like `rotZ * rotY * rotX * vec` the vector is indeed rotated "first around X then Y then Z" ?

I also think it would be better to use `RotateZYXTranslate` so that the transform can be read left to right, but since the rotation "ZYX" is to be read right-to-left anyway I think it can also be confusing

[the remark this answers is not in the document]
Comment (zbanks) [reply src/code/sys_matrix.c:300]: I think I figured out why our perspectives aren't aligned: I think it's a difference of intrinsic vs. extrinsic rotations. 
To convert between intrinsic and extrinsic rotations, reverse the order of the angles. This means "extrinsic `XYZ`" is the _same_ transformation as "intrinsic `ZYX`". 

`rotZ * rotY * rotX * vec` is the *extrinsic* rotation "first around X, then Y, then Z". However, if you perform the *intrinsic* rotation "first around Z, then Y, then X", you can derive the same value (as done in the [NASA pdf, pages 5-7](https://ntrs.nasa.gov/api/citations/19770024290/downloads/19770024290.pdf) with a diagram)

```
out = rotZ * vec''
vec'' = rotY * vec'
vec' = rotX * vec
out = rotZ * rotY * rotX * vec
```

Intrinsic rotations seem to be more common in non-graphics engineering contexts, ([e.g. aircraft](https://en.wikipedia.org/wiki/Aircraft_principal_axes), robotics). But, according to [these lecture slides, page 12](https://www.cs.utexas.edu/~theshark/courses/cs354/lectures/cs354-14.pdf), extrinsic rotations are more common in (modern-day) computer graphics. I think this is where we had a disconnect.

I think there is also an error on the Wikipedia page. There's a quote in the _[Rotation Matrix](https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix)_ section that implies that the table labels are for extrinsic rotations:
> 3. Each matrix is meant to represent, primarily, the composition of extrinsic rotations (which corresponds to the constructive evaluation of the R matrix by the multiplication of three truly elemental matrices) and, secondarily, the composition of three non elemental matrices representing globally an intrinsic rotations (around the axes of the rotating reference frame, in inverse order).

However, this line was [heavily edited after the table was added](https://en.wikipedia.org/w/index.php?title=Euler_angles&diff=782505400&oldid=765621715) -- the original text said that the table was for intrinsic rotations! :facepalm: 

---

So from here, I think we should be explicit in the function's comment that this is for intrinsic ZYX angles. 

I don't mean to cloud the discussion, but I also would be fine switching to the extrinsic convention which seems to be more popular in computer graphics. (So this function would be renamed back to `Matrix_RotateXYZ`, I guess matching the "original name," wherever that came from?)

Overall, sorry for the confusion. The nomenclature is not my strong suit, and I didn't realize that extrinsic rotations were commonly used. 
Although I like using [NASA pdf](https://ntrs.nasa.gov/api/citations/19770024290/downloads/19770024290.pdf) as a reference: it is concise, relatively complete, publicly available, and authoritative -- but it's weakness is that it doesn't cover extrinsic rotations, so maybe it was my mistake to try to introduce it here. (I do still uphold my opinion that yaw/pitch/roll should *not* be used, though :grin:)

File: src/code/sys_matrix.c [hunk 3]
```diff
@@ -740,7 +742,7 @@ void func_800D1FD4(MtxF* mf) {
 }
 
 /**
- * Gets the rotation the specified matrix represents, using Euler YXZ.
+ * Gets the rotation the specified matrix represents, using Tait-Bryan YXZ angles.
  * The flag value doesn't matter for a rotation matrix. Not 0 does extra calculation.
  */
```
Comment (zbanks) [inline src/code/sys_matrix.c:747]: It would be nice to clarify that the middle angle (`rotDest->x`) will range from `-0x4000` to `0x4000` (+/- 90 degrees), whereas the other two angles have the full +/- 180 degrees of range. 

This property is generally true of all Euler angle representations, but it might not be obvious to someone who is decomping.

File: src/code/sys_matrix.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (fig02) [reply src/code/sys_matrix.c:300]: > I guess matching the "original name," wherever that came from?
lmao, I think original in this case is actually referring to what we called it very early on, before it was renamed to RPY. So someone a while back was accidentally misleading. I cant find evidence that it was actually called that from any debug prints

[the remark this answers is not in the document]
Comment (zbanks) [reply src/code/sys_matrix.c:300]: Yeah, big :facepalm: on my part for adding to the confusion...

I'm trying to fix this error in the Wikipedia article (although we'll see if it gets reverted like Dragorn's edits did!)

[the remark this answers is not in the document]
Comment (Thar0) [reply src/code/sys_matrix.c:300]: The "original name" being `Matrix_RotateXYZ` comes from the Animal Crossing symbol map

**Comment by Dragorn421** [comment]
So, we have a choice to make here.

## Transform Matrix 101

For those who don't know, the order used to multiply matrices matters. (in general, `matA * matB != matB * matA`)

`Matrix_` functions all multiply the current matrix on the right, like `sCurrentMatrix = sCurrentMatrix * transform`

When a vector is transformed using the matrix, transforms are applied right-to-left. For example:
```c
// the vector is first rotated of a quarter turn around +y,
// then translated 10 units towards +x
// the rotation doesn't apply to the x+10 translation
Matrix_Translate(10, 0, 0, APPLY);
Matrix_RotateY(PI / 2, APPLY);
```

```c
// the vector is first translated 10 units towards +x,
// then rotated of a quarter turn around +y
// this means the x+10 translation also gets the rotation
Matrix_RotateY(PI / 2, APPLY);
Matrix_Translate(10, 0, 0, APPLY);

// (effectively, the above code is like
// rotate of a quarter turn around +y,
// then translate 10 units towards -z):
Matrix_Translate(0, 0, -10, APPLY);
Matrix_RotateY(PI / 2, APPLY);
```

For example before calling an actor's draw function the current matrix is set according to the actor's coordinates and shape rotation ([in `Actor_Draw`](https://github.com/zeldaret/oot/blob/master/src/code/z_actor.c#L2200)), and an actor may "add" (multiply) transforms onto the matrix (eg through the animation system).
Those transforms the actor adds happen before the one set in `Actor_Draw`.

## The Choice

In the above examples, I purposefully only used "elementary" matrices, that only did one operation (translation, or rotate around a single axis)
But the Matrix_ functions have more than that. For example `Matrix_RotateZYX` rotates around all three axes, and `Matrix_TranslateRotateZYX` rotates as well as translates.

With `Matrix_TranslateRotateZYX` (will leave out the `ZYX` part for now), the translation transform happens after the rotation transform.
Since it matters whether the translation or rotation happens first, **we should decide if `Matrix_TranslateRotate` should be named `Matrix_TranslateRotate` or `Matrix_RotateTranslate`**.

`Matrix_TranslateRotate` is consistent with the left-to-right matrix multiplication `sCurrentMatrix = sCurrentMatrix * translateTransform * rotateTransform`,
and also with the order individual functions would be called in to achieve the same thing: `Matrix_Translate(); Matrix_Rotate();`

`Matrix_RotateTranslate` is consistent with the transform order (a vector is rotated, then translated)

## Also

Order of rotation around different axes also matters. For example
```c
// turns +x into -z
Matrix_RotateY(PI / 2, APPLY); // turns +x into -z
Matrix_RotateX(PI / 2, APPLY); // turns +x into +x

// turns +x into +y (not -z !)
Matrix_RotateX(PI / 2, APPLY); // turns -z into +y
Matrix_RotateY(PI / 2, APPLY); // turns +x into -z
```

That's why in `Matrix_RotateZYX` the `ZYX` part matters. For example
```c
// this single call
Matrix_RotateZYX(x, y, z, APPLY);

// is equivalent to
Matrix_RotateZ(z, APPLY);
Matrix_RotateY(y, APPLY);
Matrix_RotateX(x, APPLY);
```

This is the convention used in mathematics I believe, but apparently computer graphics use more often `XYZ` to reflect the transformation order.

Similarly to as discussed above, **we need to choose either ZYX or XYZ to reflect rotation orders**.

I personally think it would be better to keep the two choices "in sync". That is, either

go full matrix-multiplication-order `Matrix_TranslateRotateZYX`
or go full transform-order `Matrix_RotateXYZTranslate`

(I would avoid something mixing the two like `Matrix_TranslateRotateXYZ` or `Matrix_RotateZYXTranslate`)

## TLDR

> we should decide if `Matrix_TranslateRotate` should be named `Matrix_TranslateRotate` or `Matrix_RotateTranslate`

> we need to choose either ZYX or XYZ to reflect rotation orders

**Comment by Dragorn421** [comment]
If I'm allowed to make an executive decision here, I think that the way the PR currently has the names is good. (if I could approve my own PR I would)

For example about naming `Matrix_TranslateRotateZYX`. In this PR:
```c
Matrix_Translate(tx, ty, tz, MTXMODE_APPLY);
Matrix_RotateZ(rz, MTXMODE_APPLY);
Matrix_RotateY(ry, MTXMODE_APPLY);
Matrix_RotateX(rx, MTXMODE_APPLY);
```
is equivalent to
```c
Vec3f translation = {tx, ty, tz};
Vec3s rotation = {rx, ry, rz};
Matrix_TranslateRotateZYX(translation, rotation);
```

Part of the reasoning for the name is:
`Matrix_Translate + Matrix_RotateZ + Matrix_RotateY + Matrix_RotateX = Matrix_TranslateRotateZYX`
when read left-to-right

**Review by zbanks — APPROVED** [review]

**Review by Roman971 — APPROVED** [review]
